quinn-proto/src/endpoint.rs

```diff
@@ -857,12 +804,218 @@ impl IndexMut<ConnectionHandle> for Slab<ConnectionMeta> {
 pub enum DatagramEvent {
     /// The datagram is redirected to its `Connection`
     ConnectionEvent(ConnectionHandle, ConnectionEvent),
-    /// The datagram has resulted in starting a new `Connection`
-    NewConnection(ConnectionHandle, Connection),
+    /// The datagram may result in starting a new `Connection`
+    NewConnection(IncomingConnection),
     /// Response generated directly by the endpoint
     Response(Transmit),
 }
 
+/// An incoming connection for which the server has not yet begun its part of the handshake.
+pub struct IncomingConnection {
+    addresses: FourTuple,
+    ecn: Option<EcnCodepoint>,
+    packet: Packet,
+    rest: Option<BytesMut>,
+    crypto: Keys,
+    src_cid: ConnectionId,
+    dst_cid: ConnectionId,
+    packet_number: u64,
+    version: u32,
+    retry_src_cid: Option<ConnectionId>,
+    orig_dst_cid: ConnectionId,
+}
+
+impl IncomingConnection {
+    /// Whether the socket address that is initiating this connection has been validated.
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive traffic
+    /// sent to `self.remote_address()`.
+    pub fn is_validated(&self) -> bool {
+        self.retry_src_cid.is_some()
+    }
+
+    /// The local IP address which was used when the peer established
+    /// the connection
+    ///
+    /// This has the same behavior as [`Connection::local_ip`]
+    pub fn local_ip(&self) -> Option<IpAddr> {
+        self.addresses.local_ip
+    }
+
+    /// The peer's UDP address.
+    pub fn remote_address(&self) -> SocketAddr {
+        self.addresses.remote
+    }
+
+    /// Whether it is legal to require the client to retry.
+    ///
+    /// If `is_validated` is false, `may_retry` is necessarily true.
+    pub fn may_retry(&self) -> bool {
+        // Currently, we do not produce retry tokens in a way that allows us to tell the difference
+        // between ones minted for a retry packet versus for a NEW_TOKEN frame, so to be safe we
+        // must simply never respond with a retry packet to an initial packet with a token in it.
+        // However, the QUIC protocol would allow for such a distinction to be implemented.
+        !self.is_validated()
+    }
+
+    /// Attempt to accept this incoming connection (an error may still occur).
+    pub fn accept(
```

Comment:
I think it would make more sense to do this the other way around, keeping an `accept()` method on `Endpoint` that takes ownership of the `IncomingConnection`.

Reply:
Done.